gammapy/maps/base.py

```diff
@@ -226,7 +226,7 @@ def from_geom(geom, meta=None, data=None, map_type="auto", unit=""):
                 raise ValueError("Unrecognized geom type.")
 
         cls_out = Map._get_map_cls(map_type)
-        return cls_out(geom, data=data, meta=meta, unit=unit)
+        return cls_out(geom, data=data, meta=meta, unit=unit, dtype=dtype)
```

Comment:
This calls `Map.__init__`? That doesn't seem to have a separate `dtype` argument. Do we really need the option to do a data dtype change here (instead of calling `astype` explicitly)? Does this work? Add test?

Reply:
Both [WcsNDMap.__init__](https://github.com/gammapy/gammapy/blob/master/gammapy/maps/wcsnd.py#L46) and [HpxNDMap.__init__](https://github.com/gammapy/gammapy/blob/master/gammapy/maps/hpxnd.py#L35) have a `dtype` argument and I guess it's needed, because you want to create the default data given a geometry. I can add a separate test for this...

Reply:
Ah, OK. I only looked at [Map.__init__](https://github.com/gammapy/gammapy/blob/bddde6ebd446be109be964f64875ba21993aa326/gammapy/maps/base.py#L35-L43) which doesn't have this.

I wonder if we can simplify the whole from_geom and copy logic for maps. It's really complex, calling up and down the class hierarchy.
One thought: would changing to `Map.from_geom` to `geom.make_map` help? This hard-coded if / else on two pre-defined types means it's non-extensible to other sub-classes, and I think generally is a bad pattern with class inheritance:
https://github.com/gammapy/gammapy/blob/a10d7c8434a04c5ddaa526d38c99d162f623700d/gammapy/maps/base.py#L191-L229

Reply:
I think API-wise the pattern is not too bad. It follows the logic of creating the map geometry
first and then creating a map by combining it with the data or crating default data. I guess `Map.from_geom()` was introduced to have a factory function, that is agnostic of the map type and I think you definitely need something like this because in the code you want to create maps, without knowing the map-type in advance. The problem of the hard-coded map sub-types you also fave for `Map.create()`.

My proposal would be to stick with the `.from_geom()`, but I agree we should change to a registry pattern for sub-map types and get rid of the hard-coded classes mappings.

Reply:
It doesn't solve all problems, but moving `Map.from_geom` to `geom.make_make` does solve this inheritance issue, allowing us to remove this if / else / ininstance check, because the geom is the concrete subclass and knows it's type. But OK, it's not clear if it's worth making the change. Let's drop this discussion for now, and then, if we have time in a few weeks, consider improving gammapy.maps.